atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java

```diff
@@ -336,11 +342,91 @@ public void hostIsAutomaticallyAddedOnRefresh() {
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+        deterministicExecutor.tick(POOL_REFRESH_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+    }
+
+    @Test
+    public void onlyValidatedHostsAreAddedOnRefresh() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+
+        setCassandraServersTo(CASS_SERVER_1);
+
+        CassandraClientPoolImpl clientPool = createClientPool();
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1);
+        setupHostsWithInconsistentTopology(ImmutableSet.of(CASS_SERVER_2));
 
         setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
         deterministicExecutor.tick(POOL_REFRESH_INTERVAL_SECONDS, TimeUnit.SECONDS);
-        assertThat(poolServers).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_3);
+        assertThat(clientPool.getAbsentHostTracker().returnPool(CASS_SERVER_2)).isPresent();
+    }
+
+    @Test
+    public void throwsWhenNoServersInPoolAndServersAddedPresentButAreInvalid() {
+        setupHostsWithInconsistentTopology(ImmutableSet.of(CASS_SERVER_1));
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+        setCassandraServersTo(CASS_SERVER_1);
+        assertThatThrownBy(this::createClientPool).isInstanceOf(SafeIllegalStateException.class);
+    }
+
+    @Test
+    public void reuseAbsentHostsContainerIfPresent() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy(), CASS_SERVER_2.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
+        createClientPool();
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2);
+        CassandraClientPoolingContainer cassandraServerTwoContainer = poolServers.get(CASS_SERVER_2);
+
+        setCassandraServersTo(CASS_SERVER_1);
+        deterministicExecutor.tick(POOL_REFRESH_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
+        deterministicExecutor.tick(POOL_REFRESH_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        assertThat(poolServers.keySet()).containsExactlyInAnyOrder(CASS_SERVER_1, CASS_SERVER_2);
+        assertThat(cassandraServerTwoContainer).isEqualTo(poolServers.get(CASS_SERVER_2));
+    }
+
+    @Test
+    public void reuseContainerIfPreviouslyInvalid() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy(), CASS_SERVER_2.proxy()));
```

Comment:
This test and the above are a bit hard to read, and I'm not _entirely_ sure I understand what property one is testing that the other isn't.

If we can't stick to arrange act assert given the complexity of the tests, either simplify or add comments s.t. the state of the world is clear.

Reply:
Also: add a test for https://github.com/palantir/atlasdb/pull/6317/files#diff-c8381f15121edbde8d40e9854428f137b054f9b222b0887bd590efefd98ccf4bR380

Reply:
Let's remove `reuseContainerIfPreviouslyInvalid`, as it's effectively a subtest of `reuseAbsentHostsContainerIfPresent`.  

`reuseContainerIfPreviouslyInvalid` primarily tested that we would use the container in the absent host tracker if it was present, where as `reuseContainerIfPreviouslyInvalid` tested that we actually store the container we generated!